It looks like there are some compile errors on this PR. The missing members can be found in t->attr

```shell
/__w/zephyr/zephyr/lib/posix/pthread.c:697:22: error: 'struct posix_thread' has no member named 'cancel_state'
  697 |                 if (t->cancel_state != PTHREAD_CANCEL_ENABLE) {
      |                      ^~
/__w/zephyr/zephyr/lib/posix/pthread.c:700:22: error: 'struct posix_thread' has no member named 'cancel_pending'
  700 |                 if (t->cancel_pending) {
      |                      ^~
/__w/zephyr/zephyr/lib/posix/pthread.c:702:26: error: 'struct posix_thread' has no member named 'cancel_state'
  702 |                         t->cancel_state = PTHREAD_CANCEL_DISABLE;
```

--- a/doc/services/portability/posix/option_groups/index.rst
+++ b/doc/services/portability/posix/option_groups/index.rst
@@ -64,7 +64,7 @@ multiple processes.
     pthread_setcanceltype(),yes
     pthread_setspecific(),yes
     pthread_sigmask(),yes
-    pthread_testcancel(),
+    pthread_testcancel(),yes
 
 .. _posix_option_group_posix_threads_ext:
 

--- a/include/zephyr/posix/pthread.h
+++ b/include/zephyr/posix/pthread.h
@@ -440,6 +440,7 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
 		   void *(*threadroutine)(void *), void *arg);
 int pthread_setcancelstate(int state, int *oldstate);
 int pthread_setcanceltype(int type, int *oldtype);
+void pthread_testcancel(void);
 int pthread_attr_setschedparam(pthread_attr_t *attr,
 			       const struct sched_param *schedparam);
 int pthread_setschedparam(pthread_t pthread, int policy,

--- a/lib/posix/options/pthread.c
+++ b/lib/posix/options/pthread.c
@@ -680,6 +680,35 @@ int pthread_setcanceltype(int type, int *oldtype)
 }
 
 /**
+ * @brief Create a cancellation point in the calling thread.
+ *
+ * See IEEE 1003.1
+ */
+void pthread_testcancel(void)
+{
+	struct posix_thread *t;
+	bool cancel_pended = false;
+
+	K_SPINLOCK(&pthread_pool_lock) {
+		t = to_posix_thread(pthread_self());
+		if (t == NULL) {
+			K_SPINLOCK_BREAK;
+		}
+		if (t->attr.cancelstate != PTHREAD_CANCEL_ENABLE) {
+			K_SPINLOCK_BREAK;
+		}
+		if (t->attr.cancelpending) {
+			cancel_pended = true;
+			t->attr.cancelstate = PTHREAD_CANCEL_DISABLE;
+		}
+	}
+
+	if (cancel_pended) {
+		posix_thread_finalize(t, PTHREAD_CANCELED);
+	}
+}
+
+/**
  * @brief Cancel execution of a thread.
  *
  * See IEEE 1003.1

--- a/tests/posix/common/src/pthread.c
+++ b/tests/posix/common/src/pthread.c
@@ -487,6 +487,46 @@ ZTEST(pthread, test_pthread_cleanup)
 	zassert_ok(pthread_join(th, NULL));
 }
 
+static bool testcancel_ignored;
+static bool testcancel_failed;
+
+static void *test_pthread_cancel_fn(void *arg)
+{
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL));
+
+	testcancel_ignored = false;
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	testcancel_ignored = true;
+
+	/* this will mark it pending */
+	zassert_ok(pthread_cancel(pthread_self()));
+
+	/* enable the thread to be cancelled */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL));
+
+	testcancel_failed = false;
+
+	/* this should terminate the thread */
+	pthread_testcancel();
+
+	testcancel_failed = true;
+
+	return NULL;
+}
+
+ZTEST(pthread, test_pthread_testcancel)
+{
+	pthread_t th;
+
+	zassert_ok(pthread_create(&th, NULL, test_pthread_cancel_fn, NULL));
+	zassert_ok(pthread_join(th, NULL));
+	zassert_true(testcancel_ignored);
+	zassert_false(testcancel_failed);
+}
+
 static void before(void *arg)
 {
 	ARG_UNUSED(arg);

--- a/tests/posix/headers/src/pthread_h.c
+++ b/tests/posix/headers/src/pthread_h.c
@@ -158,7 +158,7 @@ ZTEST(posix_headers, test_pthread_h)
 		zassert_not_null(pthread_spin_lock);
 		zassert_not_null(pthread_spin_trylock);
 		zassert_not_null(pthread_spin_unlock);
-		/* zassert_not_null(pthread_testcancel); */ /* not implemented */
+		zassert_not_null(pthread_testcancel);
 	}
 }
 #pragma GCC diagnostic pop

